test/common/http/http1/conn_pool_test.cc

```diff
@@ -380,6 +382,63 @@ TEST_F(Http1ConnPoolImplTest, ConnectFailure) {
   EXPECT_EQ(1U, cluster_->stats_.upstream_rq_pending_failure_eject_.value());
 }
 
+/**
+ * Tests that connection creation time is recorded correctly even in cases where
+ * there are multiple pending connection creation attempts to the same upstream.
+ */
+TEST_F(Http1ConnPoolImplTest, MeasureConnectTime) {
+  constexpr uint64_t sleep1_ms = 20;
+  constexpr uint64_t sleep2_ms = 10;
+  constexpr uint64_t sleep3_ms = 5;
+  Event::SimulatedTimeSystem simulated_time;
+
+  // Allow concurrent creation of 2 upstream connections.
+  cluster_->resetResourceManager(2, 1024, 1024, 1, 1);
+
+  InSequence s;
+
+  // Start the first connect attempt.
+  conn_pool_.expectClientCreate();
+  ActiveTestRequest r1(*this, 0, ActiveTestRequest::Type::Pending);
+
+  // Move time forward and start the second connect attempt.
+  simulated_time.sleep(std::chrono::milliseconds(sleep1_ms));
+  conn_pool_.expectClientCreate();
+  ActiveTestRequest r2(*this, 1, ActiveTestRequest::Type::Pending);
+
+  // Move time forward, signal that the first connect completed and verify the time to connect.
+  uint64_t upstream_cx_connect_ms1 = 0;
+  simulated_time.sleep(std::chrono::milliseconds(sleep2_ms));
+  EXPECT_CALL(*conn_pool_.test_clients_[0].connect_timer_, disableTimer());
+  EXPECT_CALL(cluster_->stats_store_,
+              deliverHistogramToSinks(Property(&Stats::Metric::name, "upstream_cx_connect_ms"), _))
+      .WillOnce(SaveArg<1>(&upstream_cx_connect_ms1));
+  r1.expectNewStream();
+  conn_pool_.test_clients_[0].connection_->raiseEvent(Network::ConnectionEvent::Connected);
+  EXPECT_EQ(sleep1_ms + sleep2_ms, upstream_cx_connect_ms1);
```

Comment:
Without changes to source/common/http/http1/conn_pool.*, this test fails in the following way:

[ RUN      ] Http1ConnPoolImplTest.MeasureConnectTime
test/common/http/http1/conn_pool_test.cc:415: Failure
Expected equality of these values:
  sleep1_ms + sleep2_ms
    Which is: 30
  upstream_cx_connect_ms1
    Which is: 10
Stack trace:
  0x43ed60: Envoy::Http::Http1::(anonymous namespace)::Http1ConnPoolImplTest_MeasureConnectTime_Test::TestBody()
  0x18aa8de: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x189af6e: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x1884df3: testing::Test::Run()
  0x18859f8: testing::TestInfo::Run()
... Google Test internal frames ...